compiler/daml-lf-tools/tests/DA/Daml/LF/Simplifier/Tests.hs

```diff
@@ -0,0 +1,118 @@
+-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Daml.LF.Simplifier.Tests
+    ( main
+    ) where
+
+import Test.Tasty
+import Test.Tasty.HUnit
+import qualified Data.NameMap as NM
+import qualified Data.Text as T
+
+import DA.Daml.LF.Ast.Base
+import DA.Daml.LF.Ast.Util
+import DA.Daml.LF.Ast.Version (versionDev)
+import DA.Daml.LF.Ast.World (initWorld)
+import DA.Daml.LF.Simplifier (simplifyModule)
+
+
+main :: IO ()
+main = defaultMain $ testGroup "DA.Daml.LF.Simplifier"
+    [ constantLiftingTests
+    ]
+
+constantLiftingTests :: TestTree
+constantLiftingTests = testGroup "Constant Lifting"
+    [ mkTestCase "empty module" [] []
+    , mkTestCase "closed value"
+        [ dval "foo" TInt64 (EBuiltin (BEInt64 10)) ]
+        [ dval "foo" TInt64 (EBuiltin (BEInt64 10)) ]
+    , mkTestCase "nested int"
+        [ dval "foo" (TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64) (EBuiltin (BEInt64 10))) ]
+        [ dval "foo" (TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64) (EBuiltin (BEInt64 10))) ]
+    , mkTestCase "nested arithmetic"
+        [ dval "foo" (TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64)
+                (EBuiltin BEAddInt64
+                    `ETmApp` EBuiltin (BEInt64 10)
+                    `ETmApp` EBuiltin (BEInt64 10)))
+        ]
+        [ dval "$sc_foo_1" TInt64
+            (EBuiltin BEAddInt64
+                `ETmApp` EBuiltin (BEInt64 10)
+                `ETmApp` EBuiltin (BEInt64 10))
+        , dval "foo" (TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64) (exprVal "$sc_foo_1"))
+        ]
+    , mkTestCase "\\xy.y" -- test that we aren't breaking up λxy.y into two lambdas.
+        [ dval "foo" (TInt64 :-> TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64)
+                (ETmLam (ExprVarName "y", TInt64)
+                    (EVar (ExprVarName "y"))))
+        ]
+        [ dval "foo" (TInt64 :-> TInt64 :-> TInt64)
+            (ETmLam (ExprVarName "x", TInt64)
+                (ETmLam (ExprVarName "y", TInt64)
+                    (EVar (ExprVarName "y"))))
+        ]
+    , mkTestCase "\\z.(\\xy.y)z" -- test that we're lifting closed lambda subexpressions
```

Comment:
Aren't we rewriting that into
```
\z. let x = z in \y. y
```
because `z` is safe?

Reply:
We are not, because `\xy.y` is closed and is therefore lifted.

It's arguable what the right thing to do is (the eta reduced `\z\y.y` would be even better than `\z. let x = z in \y. y`), but we can achieve both of these with separate optimizations (inlining vs eta reduction).